vaadin-upload-flow-parent/vaadin-upload-flow/src/main/java/com/vaadin/flow/component/upload/Upload.java

```diff
@@ -858,4 +898,62 @@ public void streamingFailed(StreamVariable.StreamingErrorEvent event) {
             }
         }
     }
+
+    /**
+     * An internal {@link TransferProgressListener} implementation to track and
+     * handle the progress of an upload operation in the component. This
+     * listener provides hooks for reacting to upload lifecycle events, such as
+     * start, progress updates, error occurrences, and successful completion.
+     * The associated Upload instance is notified at various stages of the
+     * upload to enable further handling or state updates. Doesn't need to be
+     * removed and re-added on attach/detach because this is checked on a low
+     * level in a framework.
+     */
+    private static final class UploadTransferProgressListener
+            implements TransferProgressListener {
+        private final Upload upload;
+
+        public UploadTransferProgressListener(Upload upload) {
+            this.upload = upload;
+        }
+
+        @Override
+        public void onStart(TransferContext context) {
+            upload.startUpload();
+            upload.fireStarted(context.fileName(), context.contentType(),
+                    context.totalBytes());
+        }
+
+        @Override
+        public void onProgress(TransferContext context, long transferredBytes,
+                long totalBytes) {
+            upload.fireUpdateProgress(transferredBytes, totalBytes,
+                    context.fileName());
+        }
+
+        @Override
+        public void onError(TransferContext context, IOException reason) {
+            try {
+                upload.fireUploadInterrupted(context.fileName(),
+                        context.contentType(), context.bytesTransferred(),
+                        reason);
+            } finally {
+                upload.endUpload();
+                upload.fireUploadFinish(context.fileName(),
+                        context.contentType(), context.totalBytes());
+            }
+        }
+
+        @Override
+        public void onComplete(TransferContext context, long transferredBytes) {
+            try {
+                upload.fireUploadSuccess(context.fileName(),
+                        context.contentType(), context.totalBytes());
+            } finally {
+                upload.endUpload();
+                upload.fireUploadFinish(context.fileName(),
+                        context.contentType(), context.totalBytes());
+            }
+        }
+    }
```

Comment:
I'm not sure 100% that when the `UploadHandler` is used the Upload component should fire it's own progress events. I.e. events like `StartedEvent` will be never fired if the handler is used instead of receiver.

Reply:
Component's event should be deprecated and not invoked when `UploadHandler` is used.